axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/SaleOrderServiceSupplychainImpl.java

```diff
@@ -234,4 +246,114 @@ public void updateToConfirmedStatus(SaleOrder saleOrder) {
     saleOrder.setStatusSelect(SaleOrderRepository.STATUS_ORDER_CONFIRMED);
     saleOrderRepo.save(saleOrder);
   }
+
+  @Override
+  public String createShipmentCostLine(SaleOrder saleOrder) throws AxelorException {
+    SaleOrderComputeService saleOrderComputeService = Beans.get(SaleOrderComputeService.class);
+    SaleOrderMarginService saleOrderMarginService = Beans.get(SaleOrderMarginService.class);
+    List<SaleOrderLine> saleOrderLines = saleOrder.getSaleOrderLineList();
+    Partner client = saleOrder.getClientPartner();
+    ShipmentMode shipmentMode = saleOrder.getShipmentMode();
+
+    if (shipmentMode == null) {
+      return null;
+    }
+    Product shippingCostProduct = shipmentMode.getShippingCostsProduct();
+    if (shippingCostProduct == null) {
+      return null;
+    }
+    BigDecimal carriagePaidThreshold = shipmentMode.getCarriagePaidThreshold();
+    if (client != null) {
+      List<CustomerShippingCarriagePaid> carriagePaids =
+          client.getCustomerShippingCarriagePaidList();
+      for (CustomerShippingCarriagePaid customerShippingCarriagePaid : carriagePaids) {
+        if (shipmentMode.getId() == customerShippingCarriagePaid.getShipmentMode().getId()) {
+          if (customerShippingCarriagePaid.getShippingCostsProduct() != null) {
+            shippingCostProduct = customerShippingCarriagePaid.getShippingCostsProduct();
+          }
+          carriagePaidThreshold = customerShippingCarriagePaid.getCarriagePaidThreshold();
+          break;
+        }
+      }
+    }
+    if (carriagePaidThreshold != null && shipmentMode.getHasCarriagePaidPossibility()) {
+      if (computeExTaxTotalWithoutShippingLines(saleOrder).compareTo(carriagePaidThreshold) >= 0) {
+        String message = removeShipmentCostLine(saleOrder);
+        saleOrderComputeService.computeSaleOrder(saleOrder);
+        saleOrderMarginService.computeMarginSaleOrder(saleOrder);
+        return message;
+      }
+    }
+    if (alreadyHasShippingCostLine(saleOrder, shippingCostProduct)) {
+      return null;
+    }
+    SaleOrderLine shippingCostLine = createShippingCostLine(saleOrder, shippingCostProduct);
+    saleOrderLines.add(shippingCostLine);
+    saleOrderComputeService.computeSaleOrder(saleOrder);
+    saleOrderMarginService.computeMarginSaleOrder(saleOrder);
+    return null;
+  }
+
+  private boolean alreadyHasShippingCostLine(SaleOrder saleOrder, Product shippingCostProduct) {
```

Comment:
Change all private methods to protected, as the use of private is preventing any override from customer module